src/stirling/source_connectors/socket_tracer/protocols/mongodb/stitcher.cc

```diff
@@ -0,0 +1,198 @@
+    uint64_t latest_resp_ts = 0;
+
+    // Loop until the first frame in the response deque finds a match with the oldest request that
+    // occured just before the response.
+    for (const auto& [idx, resp_frame] : Enumerate(resp_deque)) {
+      if (resp_frame.consumed) {
+        continue;
+      }
+
+      latest_resp_ts = resp_frame.timestamp_ns;
+
+      auto curr_resp = resp_frame;
+
+      // Find and insert all of the moreToCome frame(s) section data to the head response frame.
+      while (curr_resp.more_to_come) {
+        // Find the next response's deque.
+        auto next_resp_deque_it = resps->find(curr_resp.request_id);
+        if (next_resp_deque_it == resps->end()) {
+          VLOG(1) << absl::Substitute(
+              "Did not find a response deque extending the prior more to come response. "
+              "requestID: $0",
+              curr_resp.request_id);
+          error_count++;
+          break;
+        }
+
+        // Response deque containg the next more to come response frame.
+        auto& next_resp_deque = next_resp_deque_it->second;
+
+        // Find the next response frame from the deque with a timestamp just greater than the
+        // current response frame's timestamp.
+        auto next_resp_it = std::upper_bound(
+            next_resp_deque.begin(), next_resp_deque.end(), latest_resp_ts,
+            [](const uint64_t ts, const mongodb::Frame& frame) { return ts < frame.timestamp_ns; });
+        if (next_resp_it->timestamp_ns < latest_resp_ts) {
+          VLOG(1) << absl::Substitute(
+              "Did not find a response extending the prior more to come response. RequestID: $0",
+              curr_resp.request_id);
+          error_count++;
+          break;
+        }
+
+        mongodb::Frame& next_resp = *next_resp_it;
+        resp_frame.sections.insert(std::end(resp_frame.sections), std::begin(next_resp.sections),
+                                   std::end(next_resp.sections));
+        next_resp.consumed = true;
+        latest_resp_ts = next_resp.timestamp_ns;
+        curr_resp = next_resp;
+        next_resp_deque.erase(next_resp_it);
+      }
+
+      // Find the corresponding request frame for the head response frame.
+      auto req_frame_it = std::upper_bound(req_deque.begin(), req_deque.end(), latest_resp_ts,
+                                           [](const uint64_t ts, const mongodb::Frame& frame) {
+                                             return ts < frame.timestamp_ns;
+                                           }) -
+                          1;
+      if (req_frame_it == req_deque.begin() &&
+          req_frame_it->timestamp_ns > resp_frame.timestamp_ns) {
+        VLOG(1) << absl::Substitute(
+            "Did not find a request frame that is earlier than the response. Response's "
+            "responseTo: $0",
+            resp_frame.response_to);
+        resp_deque.erase(resp_deque.begin() + idx);
+        error_count++;
+        break;
+      }
```

Comment:
I was expecting this logic to come before the more to come parsing. Don't we need to identify the correct response with the timestamp heuristic before the 1 or n responses can be parsed?